src/Components/Server/src/Circuits/CircuitId.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    // Consists of a secret (data protected payload) and a non-secret identifier
+    // for use in logs and user code.
+    //
+    // The contract of this is that the id is derived from the Secret. We use the secret
+    // for comparisons, but we use the id for display (and exposing to user code). As a result,
+    // we don't include the id in any comparisons done by this class.
+    //
+    // Intentionally not overriding ToString here so that this won't accidentally
+    // get logged. It's ok to log the secret at TRACE.
+    internal readonly struct CircuitId : IEquatable<CircuitId>
+    {
+        public CircuitId(string secret, string id)
+        {
+            Secret = secret ?? throw new ArgumentNullException(nameof(secret));
+            Id = id ?? throw new ArgumentNullException(nameof(id));
+        }
+
+        public string Id { get; }
+
+        public string Secret { get; }
+
+        public bool Equals([AllowNull] CircuitId other)
+        {
+            // We want to use a fixed time equality comparison for a *real* comparisons.
+            // The only use case for Secret being null is with a default struct value,
+            // which wouldn't be the result of untrusted input.
+            if (other.Secret == null)
+            {
+                return Secret == null;
+            }
+
+            return
+                CryptographicOperations.FixedTimeEquals(
+                    MemoryMarshal.AsBytes(Secret.AsSpan()),
+                    MemoryMarshal.AsBytes(other.Secret.AsSpan()));
+        }
```

Comment:
I don't think we need to do this, given that the payload for the secret is DataProtected the ID should be by definition trusted data, and given that the Id is not a secret you can simply compare ids and avoid the unnecessary crypto. You are not giving away any information about the secret if you just do equals based on the ID

Reply:
I'm not totally convinced by this. Why would we compare just the ID?

To me it's much more correct to compare the thing that's actually used as a key (the secret).